pyquil/api/_qpu.py

```diff
@@ -197,14 +207,22 @@ def run(self, run_priority: Optional[int] = None) -> "QPU":
 
         return self
 
+    def _qpu_client(self) -> QPUClient:
```

Comment:
Technically, there's a remote possibility that the engagement expires between retrieving this client and actually using it. The client itself has no mechanism to detect this or refresh its engagement (that's in the realm of the EngagementManager).

I don't think it's a big enough problem to worry about, but there could be some instances where users get errors due to them executing a program just before any current engagement they have expires.

That said, I'm open to ideas.